Introduce an ErrorBoundary component and use it around the SnippetEditorFields

afercia opened the pull request:
## Summary

* Introduce an ErrorBoundary component to be used for the Snippet Editor.

## Relevant technical choices:

- for now, the error boundary wraps the SnippetEditorFields component
- it can be changed to wrap any other components tree, if desired
- uses `a11ySpeak` to deliver an audible message to screen reader users

Design feedback needed /Cc @hedgefield 

We can't predict where we'll need to re-use this component and show errors, so we can't set arbitrary paddings / margins. For this reason, I've aligned the text to the centre and used a bottom border, as it seemed the most pragmatical solution to me. Of course, this can be changed if desired, just let me know. Personally, I'd add `moar` white space. Screenshot:

<img width="816" alt="screen shot 2018-06-19 at 17 32 52" src="https://user-images.githubusercontent.com/1682452/41609088-4aad4a70-73ea-11e8-8b81-e23205a9f06d.png">

## Test instructions
You can test the standalone version or yarn-link to the plugin:

- with Microsoft Edge on Windows: select the title field content using Ctrl+A
- with the content still selected, press any character key to trigger the Edge error
- check the error message is correctly displayed
- an audible message is sent to the ARIA live region

If you can't test with Edge, you have to change the error boundary initial state to `true`, then open the snippet editor to see the error message.

Fixes #591

igorschoester commented:
Acceptance 🚧 

I did not get the error on Windows 10
Microsoft Edge 41.16299.492.0
Microsoft EdgeHTML 16.16299

I did notice this issue: https://github.com/Yoast/wordpress-seo/issues/10087

When putting the initial state to true on Chrome I got these errors:
<img width="1102" alt="errors" src="https://user-images.githubusercontent.com/35524806/41653244-c3bde302-7485-11e8-93a5-75ae66a9d54c.png">

afercia commented:
> When putting the initial state to true on Chrome I got these errors:

I get this error only when clicking on the preview (title, slug, or description)

I'd say it's in a way unrelated because when the error boundary renders the alternate UI (the error message), then there's no Editor rendered. As I happened to say a few times previously, a best practice is to always check the existence of an object or a DOM element before trying to use methods or get properties. In this case, `focusOnActiveFieldChange()` tries to use the `focus()` method on an element that doesn't exist at that point.

I see two ways to "fix" this:
- check for the existence of the DOM node before trying to use `activeElement.focus();` however there are probably other edge cases we may face
- wrap the error boundary around the whole UI including the snippet preview

Please do let me know what the preferred option is. /Cc @atimmer 
Also, still pending design feedback. /Cc @hedgefield

IreneStr commented on composites/basic/ErrorBoundary.js:
```diff
@@ -0,0 +1,61 @@
+	render() {
+		if ( this.state.hasError ) {
+			// Render any custom fallback UI.
+			const errorMessage = __( "Something went wrong, please reload the page.", "yoast-components" );
```
"Something went wrong. Please reload the page." 😉

IreneStr replied:
"Something went wrong. Please reload the page." 😉

afercia commented:
In the latest commits:
- wrapped the error boundary around the whole editor, this means in case of an error also the snippet preview won't be rendered
- added check for the existence of the input fields elements before trying to use `focus()` on them
- added the error boundary in the Settings pages template editors
- improved the error message text

Screenshots of how it looks in the plugin:

<img width="981" alt="screen shot 2018-06-20 at 14 58 27" src="https://user-images.githubusercontent.com/1682452/41660470-6c33d152-749c-11e8-813d-50d4f515d53f.png">

<img width="699" alt="screen shot 2018-06-20 at 14 59 06" src="https://user-images.githubusercontent.com/1682452/41660471-6c5c6a72-749c-11e8-8c86-4c4644b94d5c.png">

Xyfi commented:
CR :+1: Acceptance :+1:
